mocha/reporter.js

```diff
@@ -0,0 +1,104 @@
+const { v4: uuidv4 } = require('uuid')
+const CI = require('../util/ci')
+const Paths = require('../util/paths')
+const Mocha = require('mocha')
+const Runnable = require('mocha/lib/runnable')
+const uploadTestResults = require('../util/uploadTestResults')
+const failureExpanded = require('../util/failureExpanded')
+
+const {
+  EVENT_RUN_END,
+  EVENT_TEST_BEGIN,
+  EVENT_TEST_END,
+  EVENT_TEST_PENDING,
+} = Mocha.Runner.constants
+
+const {
+  STATE_PASSED,
+  STATE_PENDING,
+  STATE_FAILED,
+} = Runnable.constants
+
+class MochaBuildkiteAnalyticsReporter {
+  constructor(runner) {
+    this._testResults = []
+    this._testEnv = (new CI()).env();
+    this._paths = new Paths({ cwd: process.cwd() }, this._testEnv.location_prefix)
+
+    runner
+      .on(EVENT_TEST_BEGIN, (test) => {
+        this.testStarted(test)
+      })
+      .on(EVENT_TEST_END, (test) => {
+        this.testFinished(test)
+      })
+      .on(EVENT_RUN_END, () => {
+        this.testRunFinished()
+      })
+  }
+
+  testStarted(test) {
+    test.testAnalyticsId = uuidv4()
+    test.startAt = performance.now() / 1000
+  }
+
+  testFinished(test) {
+    const failureReason = test.err !== undefined ? test.err.toString() : undefined
+    const prefixedTestPath = this._paths.prefixTestPath(test.file)
+
+    this._testResults.push({
+      'id': test.testAnalyticsId,
+      'name': test.title,
+      'scope': this.scope(test),
+      'identifier': [this.scope(test), test.title].join(' ').trim(),
+      'file_name': prefixedTestPath,
+      'location': prefixedTestPath,
+      'result': this.analyticsResult(test.state),
+      'failure_reason': failureReason,
+      'failure_expanded': failureExpanded(test.err == undefined ? [] : test.err.multiple),
```

Comment:
I wasn't entirely sure what the array of hash's were supposed to be, so I followed the buildkite example and wrapped it in an expanded. https://buildkite.com/docs/test-analytics/importing-json#json-test-results-data-reference-test-result-objects

I've had a search, what is the err.multiple? Looks handy.

My method in this case was to do the following. I have used modern js though with optional chaining so that may need to be removed.
```js
  // Returns a failure expanded array of hash's
  failureExpanded(err) {
    if (!err){
      return undefined
    }

    const expanded = err?.stack?.split(/\r?\n/)
    return [{ expanded }]
  }
```

Reply:
oh, you just helped me spot a bug, thank you ... we're meant to return a `backtrace` key, not `stack, as per: https://buildkite.com/docs/test-analytics/importing-json

`err.multiple` gives you an object for each of the failures, looks like this:

```js
multiple [
  AssertionError [ERR_ASSERTION]: 41 == 42
      at Context.<anonymous> (test.js:11:12)
      at processImmediate (node:internal/timers:466:21) {
    generatedMessage: true,
    code: 'ERR_ASSERTION',
    actual: '41',
    expected: '42',
    operator: '==',
    multiple: [Array]
  }
]
```

I think we need to return an object with two keys, `expanded` and `backtrace`, I think the objects in `multiple` look appropriate for `expnded`, they just need to be converted to an array of strings to work with the API.

Then the trace should be under the `backtrace` key, I had it set to `trace` wrongly though, thanks again.